frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx

```diff
@@ -0,0 +1,74 @@
+import React, { useState, useEffect, useRef } from "react";
+import { useSelector } from "react-redux";
+import axios from "axios";
+import G6 from "@antv/g6";
+import { findDOMNode } from "react-dom";
+
+const GraphicalSelector = () => {
+  // eslint-disable-next-line
+  const { programCode, majors, minor } = useSelector((state) => state.degree);
+
+  const [courses, setCourses] = useState([]);
+  const [courseEdges, setCourseEdges] = useState([]);
+  const ref = useRef(null);
+  let graph = null;
+  const loadGraph = () => {
+    if (!graph) {
+      graph = new G6.Graph({
+        // eslint-disable-next-line react/no-find-dom-node
+        container: findDOMNode(ref.current),
+        width: 1920,
+        height: 870,
+        modes: {
+          default: ["drag-canvas", "drag-node"],
+        },
+        layout: {
+          type: "force",
+          preventOverlap: true,
+        },
+        animate: true,
+        defaultNode: {
+          size: 60,
+        },
+      });
+    }
+    const data = {
+      nodes: courses.map((c) => ({ id: c, label: c })),
+      edges: courseEdges,
+    };
+    graph.data(data);
+    graph.render();
+  };
+  // eslint-disable-next-line
+  useEffect(() => {
+    // get structure of degree
+    const fetchCourseList = async () => {
+      const res = (
+        await axios.get(`/programs/getStructure/${programCode}/${majors.join("+")}${minor && `/${minor}`}`)
+      ).data;
+      const courseList = (
+        Object.values(res.structure)
+          .flatMap((specialisation) => Object.values(specialisation)
+            .filter((spec) => typeof spec === "object" && spec.courses)
+            .flatMap((spec) => Object.keys(spec.courses)))
+      );
+      setCourses(courseList);
+      const res2 = await Promise.all(courseList.map((c) => axios.get(`/courses/getPathFrom/${c}`)));
```

Comment:
Can we pre-compute the graph structure in the BE, store that data structure in our  database and expose that through an API instead of needing to loop over every course and remake the graph on the FE.
It could save quite some computation on the FE esp for slower devices.

From what I can tell so far, this structure isnt contingent on the user planner so we could just store this data and only refresh it when it's required which could save network usage, which is probably a bigger issue than the computation issue brought up earlier.

In that case, we could store a `lastUpdated` value on both the FE and BE and then we first need to make a call to check if an update is required - which will usually return false - and then call for the data structure again only if that value is true.

Reply:
The thing i am worried about is that we may decide to use another library (like react-flow-renderer) if that turns out to be preferred by the FE team. If we solidify our structure, ill be happy to do this.

Reply:
Sounds good; we can just keep it in mind for once the FE structure gets solidified.